packages/utils/html.ts

```diff
@@ -82,3 +82,112 @@ export const extractUrls = (html: string) => {
 
 	return output;
 };
+
+export type SvgXml = {
+	title: string;
+	content: string;
+};
+
+type ExtractSvgsReturn = {
+	svgs: SvgXml[];
+	html: string;
+};
+
+export const extractSvgs = (html: string, titleGenerator: ()=> string): ExtractSvgsReturn => {
+	if (!html || !html.trim()) {
+		return {
+			svgs: [],
+			html,
+		};
+	}
+
+	const tagStack: string[] = [];
+	const body: string[] = [];
+	const svgs: SvgXml[] = [];
+	let svgStack: string[] = [];
+
+	let svgTagDepth = 0;
+	let svgStyle = '';
+
+	const currentTag = () => {
+		if (!tagStack.length) return '';
+		return tagStack[tagStack.length - 1];
+	};
+
+	const parser = new htmlparser2.Parser({
+
+		onopentag: (name: string, attrs: Record<string, string>) => {
+			tagStack.push(name);
+			attrs = { ...attrs };
+
+			if (name === 'svg') {
+				svgTagDepth++;
+				svgStyle = attrs.style || '';
+				delete attrs.style;
+				attrs.xmlns = 'http://www.w3.org/2000/svg';
+			}
+
+			let attrHtml = attributesHtml(attrs);
+			if (attrHtml) attrHtml = ` ${attrHtml}`;
+			const closingSign = isSelfClosingTag(name) ? '/>' : '>';
+			if (svgTagDepth > 0) {
+				svgStack.push(`<${name}${attrHtml}${closingSign}`);
+			} else {
+				body.push(`<${name}${attrHtml}${closingSign}`);
+			}
+		},
+
+		ontext: (decodedText: string) => {
+			if (currentTag() === 'style') {
+				if (svgTagDepth > 0) {
+					svgStack.push(decodedText.replace(/</g, '&lt;'));
+				} else {
+					body.push(decodedText.replace(/</g, '&lt;'));
+				}
+			} else {
+				if (svgTagDepth > 0) {
+					svgStack.push(decodedText.replace(/</g, '&lt;'));
+				} else {
+					body.push(htmlentities(decodedText));
+				}
+			}
+		},
+
+		onclosetag: (name: string) => {
+			if (name === 'svg') {
+				svgTagDepth--;
+				if (svgTagDepth === 0) {
+					svgStack.push('</svg>');
+					const title = `${titleGenerator()}.svg`;
+					const fullPath = `./${title}`;
+					svgs.push({
+						title,
+						content: svgStack.join(''),
+					});
+					body.push(`<img style="${svgStyle}" src="${fullPath}" />`);
+					svgStack = [];
+					svgStyle = '';
+					return;
+				}
+			}
+
+			if (svgTagDepth > 0) {
+				svgStack.push(`</${name}>`);
+			} else {
+
+				body.push(`</${name}>`);
+			}
+		},
+
+	},
+	{
+		decodeEntities: true,
+		lowerCaseAttributeNames: false,
+	});
+
+	parser.write(html);
+	parser.end();
+
+	return { svgs, html: body.join('') };
```

Comment:
What's the impact of this change on the app security? Looking at the code, I feel it will have the same SVG-related vulnerabilities that we already dealt with before. What are your thoughts on this? I guess it depends how the extracted SVGs are going to be used.

And as we've discussed before an HTML parser is not capable of parsing an SVG element safely. Maybe a different, more secure parser would be needed?